core/circuit_breaker/circuit_breaker.go

```diff
@@ -2,252 +2,520 @@ package circuit_breaker
+	// rt: the response time this entry cost.
+	handleCompleted(rt uint64, err error)
+}
+
+//================================= circuitBreakerBase ====================================
+type circuitBreakerBase struct {
+	rule                  Rule
+	retryTimeoutMs        uint64
+	nextRetryTimestamp    uint64
+	status                *Status
+	statusSwitchListeners []StatusSwitchListener
+}
+
+func (b *circuitBreakerBase) getRule() Rule {
+	return b.rule
+}
+
+func (b *circuitBreakerBase) currentStatus() Status {
+	return b.status.get()
+}
+
+// Must be overridden by concrete circuitBreaker
+func (b *circuitBreakerBase) resetMetric() {}
+
+func (b *circuitBreakerBase) retryTimeoutArrived() bool {
+	return util.CurrentTimeMillis() >= b.nextRetryTimestamp
 }
 
-// average rt circuit breaker will cut resource if the rt of resource exceed the threshold of rule.
-type averageRtCircuitBreaker struct {
-	// status of the circuit breaker
-	cut util.AtomicBool
-	// the count of request exceed the threshold
-	passCount int64
-	rule      *averageRtRule
-	metric    base.ReadStat
+func (b *circuitBreakerBase) updateNextRetryTimestamp() {
+	b.nextRetryTimestamp = util.CurrentTimeMillis() + b.retryTimeoutMs
 }
 
-func newAverageRtCircuitBreaker(rule *averageRtRule) *averageRtCircuitBreaker {
-	resNode := stat.GetResourceNode(rule.Resource)
-	var metric base.ReadStat
-	// TODO need to optimize, we should to handle the scenario that resNode is nil
-	if resNode != nil {
-		metric = resNode.GetOrCreateSlidingWindowMetric(rule.SampleCount, rule.IntervalInMs)
+// fromClosedToOpen update circuit breaker status machine from closed to open
+// Used for opening circuit breaker from closed when checking circuit breaker
+// return true if succeed to update
+func (b *circuitBreakerBase) fromClosedToOpen() bool {
+	if b.status.casStatus(Closed, Open) {
+		b.updateNextRetryTimestamp()
+		for _, listener := range b.statusSwitchListeners {
+			listener.onSwitchToOpen(Closed, b.rule)
+		}
+		return true
 	}
-	return &averageRtCircuitBreaker{
-		rule:   rule,
-		metric: metric,
+	return false
+}
+
+// fromOpenToHalfOpen update circuit breaker status machine from open to half-open
+// Used for probing
+// return true if succeed to update
+func (b *circuitBreakerBase) fromOpenToHalfOpen() bool {
+	if b.status.casStatus(Open, HalfOpen) {
+		for _, listener := range b.statusSwitchListeners {
+			listener.onSwitchToHalfOpen(Open, b.rule)
+		}
+		return true
 	}
+	return false
 }
 
-// For test
-func newAverageRtCircuitBreakerWithMetric(rule *averageRtRule, metric base.ReadStat) *averageRtCircuitBreaker {
-	return &averageRtCircuitBreaker{
-		rule:   rule,
-		metric: metric,
+// fromHalfOpenToOpen update circuit breaker status machine from half-open to open
+// Used for failing to probe
+// return true if succeed to update
+func (b *circuitBreakerBase) fromHalfOpenToOpen() bool {
+	if b.status.casStatus(HalfOpen, Open) {
+		for _, listener := range b.statusSwitchListeners {
```

Comment:
`updateNextRetryTimestamp` is also needed here.

Reply:
make sense.